keps/draft-20180426-namespace-template.md

```diff
@@ -0,0 +1,513 @@
+---
+kep-number: draft-20180426
+title: Namespace Population
+authors:
+  - "@easeway"
+owning-sig: sig-auth
+participating-sigs:
+  - sig-api-machinery
+  - sig-cluster-lifecycle
+reviewers:
+  - "@davidopp"
+  - "@tallclair"
+  - "@ericchiang"
+  - "@liggitt"
+  - "@roberthbailey"  
+approvers:
+  - "@tallclair"
+  - "@ericchiang"
+  - "@liggitt"
+  - "@roberthbailey"
+editor: "@easeway"
+creation-date: 2018-04-26
+status: provisional
+---
+
+# Namespace Population
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [User Stories](#user-stories)
+      * [Security Defaults](#security-defaults)
+    * [Implementation Details](#implementation-details)
+      * [The Controller](#the-controller)
+      * [Namespace Match](#namespace-match)
+      * [Apply the template](#apply-the-template)
+      * [Schema validation](#schema-validation)
+      * [Opt-out](#opt-out)
+      * [Self-serviced namespace creation](#self-serviced-namespace-creation)
+      * [Readiness of a namespace](#readiness-of-a-namespace)
+    * [Security Consideration](#security-consideration)
+      * [Opt-out is privileged](#opt-out-is-privileged)
+      * [Race Condition](#race-condition)
+      * [Manual Update of objects](#manual-update-of-objects)
+* [Graduation Criteria](#graduation-criteria)
+
+## Summary
+
+Namespace Population is an automated mechanism to make sure the predefined policy objects
+(e.g. NetworkPolicy, Role, RoleBinding) are present in selected namespaces.
+
+## Motivation
+
+Kubernetes users create namespaces (self-service namespace creation) and want to
+ensure security/isolation policies are enforced between them and other users.
+Just like addon manager creates cluster-scope policy objects like PodSecurityPolicy,
+ClusterRole, ClusterRoleBinding at cluster creation time.
+It creates namespace-scope policy objects during namespace creation time.
+
+### Goals
+
+The namespace population mechanism described in this proposal is effective to
+populate identical configurations into large number of namespaces.
+It's also effective to populate different sets of predefined configurations into
+different sets of namespaces.
+The mechanism is agnostic to object types as long as they are namespace scoped.
+The mechanism works in a single cluster.
+
+### Non-Goals
```

Comment:
I think you need to add as a non-goal (because this solution can't provide it)

* Defend against namespace exhaustion or malicious namespace creators.

and any elements of the proposed solution that overlap should be dropped.  Instead it is a defaulting solution that could be composed with a future change that defends namespace creation itself via some pattern or hook.

Reply:
Self-service namespace creation is a much harder problem than this proposal.